alg/teca_gradient.cxx

```diff
@@ -0,0 +1,464 @@
+#include "teca_gradient.h"
+
+#include "teca_cartesian_mesh.h"
+#include "teca_array_collection.h"
+#include "teca_variant_array.h"
+#include "teca_variant_array_impl.h"
+#include "teca_metadata.h"
+#include "teca_array_attributes.h"
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <cmath>
+
+#if defined(TECA_HAS_BOOST)
+#include <boost/program_options.hpp>
+#endif
+
+using std::string;
+using std::vector;
+using std::cerr;
+using std::endl;
+using std::cos;
+
+// #define TECA_DEBUG
+
+namespace {
+
+template <typename num_t>
+constexpr num_t deg_to_rad() { return num_t(M_PI)/num_t(180); }
+
+template <typename num_t>
+constexpr num_t earth_radius() { return num_t(6371.0e3); }
+
+// compute the gradient using second-order, centered finite difference
+// assumes fixed mesh spacing. here we add periodic bc in lon
+// and apply unit stride vector optimization strategy to loops
+template <typename num_t, typename pt_t>
+void gradient(num_t *grad_x, num_t *grad_y, const pt_t *lon, const pt_t *lat,
+    const num_t *f, unsigned long n_lon,
+    unsigned long n_lat, bool periodic_lon=true)
+{
+    size_t n_bytes = n_lat*sizeof(num_t);
+    num_t *delta_x = static_cast<num_t*>(malloc(n_bytes));
+
+    // delta lon as a function of latitude
+    num_t delta_x_equator = num_t(2)*(lon[1] - lon[0]) * deg_to_rad<num_t>() * earth_radius<num_t>();
+    for (unsigned long j = 0; j < n_lat; ++j)
+        delta_x[j] = delta_x_equator * cos(lat[j] * deg_to_rad<num_t>());
+
+    // delta lat
+    num_t delta_y = (lat[1] - lat[0]) * deg_to_rad<num_t>() * earth_radius<num_t>();
+    num_t dy = num_t(2)*delta_y;
+
+    unsigned long max_i = n_lon - 1;
+    unsigned long max_j = n_lat - 1;
+
+    // gradient
+    for (unsigned long j = 1; j < max_j; ++j)
+    {
+        unsigned long jj = j*n_lon;
+        const num_t *f_y2 = f + jj + n_lon;
+        const num_t *f_y0 = f + jj - n_lon;
+        const num_t *f_x2 = f + jj + 1;
+        const num_t *f_x0 = f + jj - 1;
+        num_t *gx = grad_x + jj;
+        num_t *gy = grad_y + jj;
+        num_t dx = delta_x[j];
+
+        for (unsigned long i = 1; i < max_i; ++i)
+        {
+            gx[i] = (f_x2[i] - f_x0[i]) / dx;
+            gy[i] = (f_y2[i] - f_y0[i]) / dy;
+        }
+    }
+
+    if (periodic_lon)
+    {
+        // periodic in longitude
+        for (unsigned long j = 1; j < max_j; ++j)
+        {
+            unsigned long jj = j*n_lon;
+            const num_t *f_y2 = f + jj + n_lon;
+            const num_t *f_y0 = f + jj - n_lon;
+            const num_t *f_x2 = f + jj + 1;
+            const num_t *f_x0 = f + jj + max_i;
+            num_t *gx = grad_x + jj;
+            num_t *gy = grad_y + jj;
+            num_t dx = delta_x[j];
+
+            gx[0] = (f_x2[0] - f_x0[0]) / dx;
+            gy[0] = (f_y2[0] - f_y0[0]) / dx;
```

Comment:
Wouldn't it be dy instead of dx in line 92?

Reply:
yes, thank you on this and the one below!!  The test wouldn't have caught this because dx=dy

Reply:
this has been fixed, thanks!